pkg/util/volid.go

```diff
@@ -60,20 +71,39 @@ Version 1 of the encoding scheme is as follows,
 	[clusterID:36bytes (MAX)] + [-:1byte]
 	[poolID:16bytes] + [-:1byte]
 	[ObjectUUID:36bytes]
-
 	Total of constant field lengths, including '-' field separators would hence be,
 	4+1+4+1+1+16+1+36 = 64
 */
+
+/*
+ComposeCSIID composes a CSI ID from passed in parameters.
+Version 2 of the encoding scheme is as follows,
+	[csi_id_version=1:4byte] + [-:1byte]
+	[length of clusterID=1:4byte] + [-:1byte]
+	[clusterID:36bytes (MAX)] + [-:1byte]
+	[poolID:16bytes] + [-:1byte]
+	[length of volNamePrefix=1:4byte] + [-:1byte]
+	[volNamePrefix:36bytes (MAX)] + [-:1byte]
+	[ObjectUUID:36bytes]
+
+	Total of constant field lengths, including '-' field separators would hence be,
+	4+1+4+1+4+1+1+16+1+36 = 69
+*/
 const (
-	knownFieldSize = 64
-	uuidSize       = 36
+	knownV1FieldSize = 64
+	knownV2FieldSize = 69
+	uuidSize         = 36
+	volNamePrefix    = "volumenameprefix"
+	snapNamePrefix   = "snapshotnameprefix"
 )
 
+// ComposeCSIID composes a CSI ID from passed in parameters
 func (ci CSIIdentifier) ComposeCSIID() (string, error) {
 	buf16 := make([]byte, 2)
 	buf64 := make([]byte, 8)
 
-	if (knownFieldSize + len(ci.ClusterID)) > maxVolIDLen {
+	ci.EncodingVersion = EncodingV2
```

Comment:
Nit: should you default to V1 unless a custom naming prefix is used? I am thinking about rolling updates when now it would be possible for new V2 names to be created and attachers might not be able to decode them.

Reply:
ComposeCSIID will be called only for the new volumes. I added code to support both V1 and V2 volumes decoding for mounting and deleting.

Reply:
Understood -- but you need to support rolling upgrades. Like I said, if you deploy a new provisioner before you fully deploy your node attacher, you are broken.

Reply:
my assumption is that there should not be any ongoing PVC creation or Binding during the upgrade procedure.

@humblec is having more idea I think he can answer it better.

Reply:
That doesn't make sense to me ... these are going to be potentially large deployments and we should *never* require the end user to shut down k8s usage while you upgrade things. That is just not how distributed systems are built and maintained.